File: rhizome/node/mobile_node.py
"""
Mobile-узел (максимально легкий клиент)
"""

from rhizome.config import Config
from rhizome.node.base_node import BaseNode


class MobileNode(BaseNode):
    """Mobile-узел для мобильных устройств"""
    
    def __init__(self, config: Config):
        # Принудительно устанавливаем тип mobile
        config.node.node_type = "mobile"
        super().__init__(config)
        
        # Сильно ограничиваем ресурсы для mobile-узла
        self.config.storage.max_storage_size = min(
            self.config.storage.max_storage_size,
            100 * 1024 * 1024  # 100 MB максимум
        )
        
        # Упрощенный протокол
        self.config.dht.k = 10  # Меньше узлов в бакетах

File: src/node/mobile_node.rs
use crate::config::Config;
use crate::node::base_node::BaseNode;
use std::ops::Deref;

/// Mobile-node for mobile device
pub struct MobileNode {
    // Используем композицию
    pub base: BaseNode,
}

#[allow(dead_code)]
impl MobileNode {
    /// Constructor for node of full type
    ///
    /// Guarantied that node type is mobile, max storage is 100mb and max buckets count is 10.
    pub async fn new(mut config: Config) -> Result<Self, Box<dyn std::error::Error>> {
        config.node.node_type = "mobile".to_string();

        let max_mobile_bytes: u64 = 100 * 1024 * 1024;
        config.storage.max_storage_size = config.storage.max_storage_size.min(max_mobile_bytes);

        config.dht.k = 10;

        let base = BaseNode::new(config).await?;

        Ok(Self { base })
    }
}

/// Realization of Deref gives opportunity to call BaseNode methods in MobileNode
/// Example: mobile_node.start() <-- mobile_node.base.start()
impl Deref for MobileNode {
    type Target = BaseNode;

    fn deref(&self) -> &Self::Target {
        &self.base
    }
}
